### ISSUE TITLE
Support for torch._dynamo.export for Phi3

### ISSUE THREAD

**Issue by Giuseppe5** [huggingface/transformers#34046]
### Feature request

Compared to `symbolic_trace`, the new (but I assume, experimental) entrypoint in `torch._dynamo.export` seems to provide a more robust way to extract modular FX graphs, that can't have any graph breaks.
I have been experimenting with some networks (Pythia, OPT, Llama, Mistral), and they all go through.

It seems that Phi3 breaks because of this line:
https://github.com/huggingface/transformers/blob/36d410dab637c133f1bb706779c75d9021d403cf/src/transformers/models/phi3/modeling_phi3.py#L213

Where `self.inv_freq` is redefined at runtime in the forward pass.
This is a bit confusing, and I would recommend to drop `self` and use a `normal` runtime variable.
I'm not sure if this has potential side effects.

A similar patter seems to be repeated in other Embedding classes in Phi3.

To reproduce:
```python
model = AutoModelForCausalLM.from_pretrained("microsoft/Phi-3.5-mini-instruct")
model, guards = torch._dynamo.export(model)(**model.dummy_inputs)
```

@gante @ArthurZucker 

### Motivation

Dropping the reference to `self.inv_freq` would allow to obtain a fullgraph with dynamo.
Having full FX graph is also a requirement for torch.export, although I have not tested that API.

### Your contribution

I can't directly contribute with a PR at the moment.
I could test a PR from my side to check compatibility with dynamo and potential side effects, once the PR is open.

**Comment by ArthurZucker** [comment]
Hey! Would you like to open a new PR for this? 🤗

**Comment by Giuseppe5** [comment]
Yes, I will look into it and update this once it's done!

**Comment by Giuseppe5** [comment]
Unfortunately I did not get the chance to work on this, and now after some refactoring, the proposed fix does not apply anymore.

However, due to the current implementation of `longrope`, Phi is still not compatible with dynamo.

In the new implementation, I'm not sure of how to take care of this problem. 
Not sure if someone could look into this.

**Comment by qgallouedec** [comment]
Hey @Giuseppe5, sorry for the slow follow-up, we're catching up on the backlog.

Two parts here.

1. Your original diagnosis is now outdated: after the RoPE refactor, `inv_freq` is a non-persistent buffer set once in `__init__` 
https://github.com/huggingface/transformers/blob/150eb7c9ed4091294c829fa0e9466b090cb0f87f/src/transformers/models/phi3/modeling_phi3.py#L83-L84
rather than being reassigned every forward, so that specific line is gone.

2. The remaining `longrope` incompatibility you raise was investigated in depth in #35716 and closed as a known limitation: the graph break comes from the data-dependent branch on `torch.max(position_ids)` in `longrope_frequency_update` https://github.com/huggingface/transformers/blob/150eb7c9ed4091294c829fa0e9466b090cb0f87f/src/transformers/modeling_rope_utils.py#L47-L80
which we can't drop without a heavy modeling change. And very few checkpoints use longrope. We can't infer seq length from cache shape instead, since that breaks with padding and static cache. You can still `torch.compile` the model (just not with `fullgraph=True`); the graph break on the rotary is tiny and shouldn't meaningfully hurt throughput.

Closing in favor of #35716, but happy to revisit if longrope adoption grows or if you find a way to make that branch export-safe without touching the hot path. 🤗
